evm_arithmetization/src/verifier.rs

```diff
@@ -133,6 +204,23 @@ where
 
     let public_values = all_proof.public_values;
 
+    // Verify shift table and kernel code.
+    for (hash1, hash2) in initial_memory_merkle_cap::<F, C, D>(
+        config.fri_config.rate_bits,
+        config.fri_config.cap_height,
+    )
+    .0
+    .iter()
+    .zip(public_values.mem_before.mem_cap.iter())
+    {
+        for (&limb1, limb2) in hash1.to_vec().iter().zip(hash2) {
+            ensure!(
+                limb1 == F::from_canonical_u64(limb2.as_u64()),
+                anyhow::Error::msg("Invalid initial MemBefore Merkle cap.")
+            );
+        }
+    }
+
```

Comment:
minor, but may be better to be in its own helper method to not clog the main `verify` function